packages/central-server/src/database/MeditrakSyncQueue.js

```diff
@@ -0,0 +1,158 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2022 Beyond Essential Systems Pty Ltd
+ */
+
+import winston from 'winston';
+import { ChangeHandler } from '@tupaia/database';
+import { createPermissionsBasedMeditrakSyncQueue } from './createPermissionsBasedMeditrakSyncQueue';
+
+const modelValidator = model => {
+  if (!model.meditrakConfig.minAppVersion) {
+    throw new Error(
+      `Model for ${model.databaseType} must have a meditrakConfig.minAppVersion property`,
+    );
+  }
+  return true;
+};
+
+const stripDataFromChange = change => {
+  // eslint-disable-next-line @typescript-eslint/naming-convention
+  // eslint-disable-next-line camelcase
+  const { old_record, new_record, ...restOfRecord } = change;
+  return { ...restOfRecord };
+};
+
+const arraysAreSame = (arr1, arr2) =>
+  arr1.length === arr2.length && arr1.every(item => arr2.includes(item));
+
+/**
+ * Adds server side changes to the meditrakSyncQueue
+ */
+export class MeditrakSyncQueue extends ChangeHandler {
+  constructor(models) {
+    super(models);
+    this.syncQueueModel = models.meditrakSyncQueue;
+
+    const typesToSync = models.getTypesToSyncWithMeditrak();
+    const modelNamesToSync = Object.entries(models)
+      .filter(([, model]) => typesToSync.includes(model.databaseType))
+      .map(([modelName]) => modelName);
+    const modelsToSync = typesToSync.map(type => models.getModelForDatabaseType(type));
+    modelsToSync.forEach(model => modelValidator(model));
+    this.changeTranslators = Object.fromEntries(
+      modelNamesToSync.map(model => [model, change => [stripDataFromChange(change)]]),
+    );
+
+    // Keep all the data for surveys, as we want to check if the permissions have changed
+    this.changeTranslators.survey = change => [change];
+  }
+
+  async createPermissionsBasedView() {
+    await createPermissionsBasedMeditrakSyncQueue(this.models.database);
+    winston.info(`Created permissions_based_meditrak_sync_queue`);
+  }
+
+  async refreshPermissionsBasedView() {
+    try {
+      const start = Date.now();
+      await this.models.database.executeSql(
+        `REFRESH MATERIALIZED VIEW CONCURRENTLY permissions_based_meditrak_sync_queue;`,
+      );
+      const end = Date.now();
+      winston.info(`permissions_based_meditrak_sync_queue refresh took: ${end - start}ms`);
```

Comment:
How long does this generally take, out of interest?

Reply:
I'll double check this, but I think I remember it taking a few seconds. Since it's a standard materialized view it basically has to fully rebuild each time, but the underlying query doesn't take too long.

If it were a fast refresh materialized view however... 😉

Reply:
Hahaha nooo don't do it!!

Reply:
Okay, so it takes 30 seconds to refresh, my bad. Makes sense as that's how long it takes to build the table initially. However, it's a[ concurrent refresh](https://www.postgresql.org/docs/current/sql-refreshmaterializedview.html), which means the table can be safely queried while it's refreshing